### PR TITLE
DScanner: Add a minimal .dscanner.ini + start to fix easy DScanner issues

### PR DESCRIPTION
**Pull request by wilzbach**
- redundant attributes
- redundant if/else
- empty declarations (required as libdparse's parser is a bit more strict.)
- misleading static if, else
 - equivalent assignment
- `const` on `opEquals` or `toHash`
- if constraint indentation

CC @BBasile

### CODE REVIEW

File: src/.dscanner.ini [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+; Process:
+; - Pick your favorite check
+; - Remove a module from the blacklist
+; - Run DScanner
+; - Fix the warnings
+; - Submit a PR to DMD
+; - GOTO: Remove a module
+;
+; Some checks are currently disabled.
+[analysis.config.ModuleFilters]
```
Comment (wilzbach) [inline src/.dscanner.ini]: This is just a very simplistic blacklist. For Phobos, we generated the entire blacklist for important checks.
See: https://github.com/dlang/phobos/pull/5501

File: src/dmd/statementsem.d [hunk 2]
```diff
@@ -915,32 +915,35 @@ private extern (C++) final class StatementSemanticVisitor : Visitor
                     return returnEarly();
                 }
             }
-            else if (!needExpansion)
```
Comment (BBasile) [inline src/dmd/statementsem.d]: The changes between here and line 946 seem unrelated to the PR.

**Review by ghost — APPROVED** [review]

File: src/posix.mak [hunk 3]
```diff
@@ -593,6 +596,28 @@ checkwhitespace: $(HOST_DMD_PATH) $(TOOLS_DIR)/checkwhitespace.d
...
+$(DSCANNER_DIR)/dsc: | $(DSCANNER_DIR)
+	# debug build is faster, but disable 'missing import' messages (missing core from druntime)
+	sed 's/dparse_verbose/StdLoggerDisableWarning/' $(DSCANNER_DIR)/makefile > $(DSCANNER_DIR)/dscanner_makefile_tmp
+	mv $(DSCANNER_DIR)/dscanner_makefile_tmp $(DSCANNER_DIR)/makefile
+	DC=$(HOST_DMD_PATH) $(MAKE) -C $(DSCANNER_DIR) githash debug
+
+# runs static code analysis with Dscanner
+dscanner: $(DSCANNER_DIR)/dsc
+	@echo "Running DScanner"
+	$(DSCANNER_DIR)/dsc --config .dscanner.ini --styleCheck $$(find dmd -name "*.d" | grep -v "func.d") -I.
```
Comment (wilzbach) [inline src/posix.mak]: Exclusion of `func.d` was necessary as libdparse currently isn't able to parse.
See https://github.com/dlang-community/libdparse/issues/258

Comment (BBasile) [reply src/posix.mak]: This can be removed already. Just update the submodule in D-Scanner and the hash in this PR.

Comment (BBasile) [reply src/posix.mak]: new hash : 3a859d39c4b59822b1bc0452b3ddcd598ef390a2

**Review by jacob-carlborg — APPROVED** [review]

File: src/posix.mak (hunk 3 above)
Comment (wilzbach) [reply src/posix.mak]: Thanks!
